src/components/action-bar/action-bar.tsx

```diff
@@ -0,0 +1,275 @@
+ * frequently used functionalities, enabling users to perform common tasks efficiently.
+ * It enhances usability by organizing important actions in a visually prominent and easily accessible location.
+ *
+ * The action bar's design and layout can vary based on the platform or application,
+ * but its primary goal remains consistent—to
+ * empower users to interact with the software and perform desired actions effortlessly.
+ * @exampleComponent limel-example-action-bar
+ * @exampleComponent limel-example-action-bar-overflow-menu
+ * @exampleComponent limel-example-action-bar-colors
+ * @exampleComponent limel-example-action-bar-floating
+ * @exampleComponent limel-example-action-bar-styling
+ * @exampleComponent limel-example-action-bar-as-primary-component
+ */
+@Component({
+    tag: 'limel-action-bar',
+    shadow: true,
+    styleUrl: 'action-bar.scss',
+})
+export class ActionBar {
+    /**
+     * Items that are placed in the action bar.
+     * These represent primary actions.
+     */
+    @Prop()
+    public actionBarItems: Array<ActionBarItem | ListSeparator> = [];
+
+    /**
+     * A label used to describe the purpose of the element to users
+     * of assistive technologies, like screen readers.
+     * Example value: "toolbar"
+     */
+    @Prop({ reflect: true })
+    public accessibleLabel?: string;
+
+    /**
+     * - When set to `fullWidth`, the component will take the
+     * entire width of its container.
+     * - When set to `floating`, the component will get basic stylings
+     * to visualize the floating state.
+     * :::note
+     * You should still properly position the component
+     * according to the structure of your user interface.
+     * For example, use an `absolute` or `fixed` position.
+     * :::
+     */
+    @Prop({ reflect: true })
+    public layout?: 'fullWidth' | 'floating';
+
+    /**
+     * Fired when a action bar item has been clicked.
+     */
+    @Event()
+    public itemSelected: EventEmitter<ActionBarItem>;
+
+    @Element()
+    public element: HTMLElement;
+
+    /**
+     * Defines the location that the content of the overflow menu
+     * appears, in relation to its trigger.
+     */
+    @Prop({ reflect: true })
+    public openDirection: OpenDirection;
+
+    @State()
+    private overflowCutoff: number = this.actionBarItems.length;
+
+    @State()
+    private calculatingOverflow: boolean = false;
+
+    private resizeObserver: ResizeObserver;
+    private intersectionObserver: IntersectionObserver;
+    private overflowRecalculationScheduled: boolean;
+    // eslint-disable-next-line no-magic-numbers
+    private RESIZE_RATE_MS = 50;
+    // eslint-disable-next-line no-magic-numbers
+    private RESIZE_TIMEOUT = 250;
+    private lastOverflowCutoff: number = this.actionBarItems.length;
+    private lastResize = Date.now();
+
+    public componentWillLoad() {
+        this.resizeObserver = new ResizeObserver(this.handleResize);
```

Comment:
If the component is disconnected from the DOM and then connected again, the resize observer will not listen for any more events. This should be moved to `connectedCallback` instead

Reply:
Or, since it also relies on the parent element, we can conditionally create it in `componentDidRender` if it has not already been created

Reply:
I'm also curious about why we need both `ResizeObserver` and `IntersectionObserver` :thinking: 
Would it not be enough to create menu items for icons that are not visible? If the container is resized so that a new item would be visible or hidden, would we not get a callback from the `IntersectionObserver` so that we can add/remove items from the menu? :thinking: I'm probably missing someting :smile:

Reply:
I think that would work, but you might get visible re-renders as you resize. Say you make it smaller, it rerenders to fit the items but also suddenly adds the menu item at the end, which suddenly falls slightly outside of the container. So it requires another render. But now there'll be a visible "glitching" for a split second.

Worst case scenario is if we have this happen near a separator, it might take 3 rerenders.

We also couldn't wait to render until someone has stopped resizing before we "reveal" the result, but this is more of an optional feature (I quite like it though)

But I might be wrong as well there 😅 , it's hard to be sure of anything with this overflow business without explicitly trying

Reply:
Or no ignore that first and third paragraph, it seems to me you could possibly mitigate that. But the second paragraph still stands 😅

Reply:
Wait I'll backpeddle my last comment again, we might need it for when we drag the container to become bigger, because the onResize handler sets the cutoff point to the number of items, which needs to happen for the algorithm to work (the algorithm cuts down on the number of items till things fit)

Reply:
I think my point is that we don't need to calculate if anything fits at all :smile: Right now it seems that we render the icons that fits and then those that do not fit inside the menu. With this approach I'm not sure why we need the `IntersectionObserver`.
On the other hand, if we would render _all_ icons _all_ the time, the `IntersectionObserver` will tell us which items that are not visible, and we can then render "copies" of those items inside the menu. If the container is resized, we will again get notified by the `IntersectionObserver` if something would be come visible or hidden, so we can adjust the menu accordingly. In this scenario, there should be no need to use `ResizeObserver`. 

Again, I might be missing something, but it seems that we should use either `ResizeObserver` or `IntersectionObserver`, not both. `IntersectionObserver` is probably the better option, so we do not have to calculate any sizes ourselves.

Reply:
I also had the idea that you might not need intersection observer in addition to resize observer. If you have the parent width, I guess it's possible to check the offset position relative to the parent somehow. Maybe you can also cache the offset of items, so you don't have to render overflowing items repeatedly?

Reply:
After talking to Johan we might be able solve this only using the intersection observer. For that to work we need to rearrange how we render the menu a bit. Now we always place it in the end of all the `action-bar-items`. But if we are going to render the ones that are invisible as well we need to do it slightly different I think.

Now both the menu and the action bar items are rendered at the same level in the DOM and the host is the element we are intersecting on.

```html
<Host>
    <action-bar-item />
    <action-bar-item />
    <action-bar-item />
    <action-bar-menu />
</Host>
```

Could we do it something like this instead?

```html
<Host>
    <div class='div-to-observe'>
        <action-bar-item />
        <action-bar-item />
        <action-bar-item />
    </div>
    <action-bar-menu />
</Host>

If we don't need to render the action bar menu that space will be claimed by the div that we observe intersections on. I don't know, just shower thoughts. But if we don't do it somehting in line with this it might get tricky to calculate were to put the menu when it's needed.